runtime/lua/vim/lsp/document_highlight.lua

```diff
@@ -0,0 +1,502 @@
+    local bufstate = bufstates[bufnr]
+    if not bufstate then
+      return
+    end
+
+    if bufstate.version == bufstate.applied then
+      return
+    end
+
+    local enabled = bufstate.enabled
+    if enabled == nil then
+      enabled = globalstate.enabled
+    end
+
+    if not enabled then
+      return
+    end
+
+    local client_highlights = bufstate.client_highlights
+
+    for row = toprow, botrow do
+      api.nvim_buf_clear_namespace(bufnr, namespace, row, row + 1)
+
+      for _, row_highlights in pairs(client_highlights) do
+        local highlights = row_highlights[row] or {}
+
+        for _, highlight in pairs(highlights) do
+          local col = highlight.range['start'].character
+          local end_row = highlight.range['end'].line
+          local end_col = highlight.range['end'].character
+
+          api.nvim_buf_set_extmark(bufnr, namespace, row, col, {
+            end_row = end_row,
+            end_col = end_col,
+            ephemeral = false,
+            hl_group = hl_group_from_kind(highlight.kind),
+          })
+        end
+      end
+    end
+
+    bufstate.applied = bufstate.version
+  end,
+})
+
+---Optional filters |kwargs|, or `nil` for all.
+---@class vim.lsp.document_highlight.enable.Filter
+---@inlinedoc
+---
+---Buffer number, or 0 for current buffer, or nil for all.
+---@field bufnr? integer
+
+---Query whether document highlight is enabled in the {filter}ed scope
+---@param filter? vim.lsp.document_highlight.enable.Filter
+---@return boolean
+function M.is_enabled(filter)
+  vim.validate({ filter = { filter, 'table', true } })
+  filter = filter or {}
+
+  local bufnr = filter.bufnr
+  if bufnr == nil then
+    return globalstate.enabled
+  end
+
+  bufnr = bufnr == 0 and api.nvim_get_current_buf() or bufnr
+  local bufstate = bufstates[bufnr]
+  if not bufstate then
+    return false
+  end
+
+  if bufstate.enabled == nil then
+    return globalstate.enabled
+  else
+    return bufstate.enabled
+  end
+end
+
+---Enables or disables document highlights for the {filter}ed scope.
+---
+---To "toggle", pass the inverse of `is_enabled()`:
+---
+---```lua
+---vim.lsp.document_highlight.enable(not vim.lsp.document_highlight.is_enabled())
+---```
+---@param enable? boolean
+---@param filter? vim.lsp.document_highlight.enable.Filter
+function M.enable(enable, filter)
+  vim.validate('enable', enable, 'boolean', true)
+  vim.validate('filter', filter, 'table', true)
+  enable = enable == nil or enable
+  filter = filter or {}
+
+  local bufnr = filter.bufnr
+  if bufnr == nil then
+    globalstate.enabled = enable
+    for b, bufstate in pairs(bufstates) do
+      bufstate.enabled = nil
+      refresh(b)
+    end
+  else
+    bufnr = bufnr == 0 and api.nvim_get_current_buf() or bufnr
+    local bufstate = bufstates[bufnr]
+    if not bufstate then
+      return
+    end
+
+    if enable == globalstate.enabled then
+      bufstate.enabled = nil
+    else
+      bufstate.enabled = enable
+    end
```

Comment:
```suggestion
    bufstate.enabled = enable ~= globalstate.enabled and enable or nil
```

Reply:
This is not correct. `and ... or` is not equivalent to the regular ternary operation. In `condition and expr1 or expr2`, if the value of `expr1` can be `nil` or `false`, regardless of the result of `condition`, the obtained value will always be `expr2`. This is what I deliberately want to avoid.

Reply:
nil trap should be avoided. enable always has a boolean available here `vim.F.if_nil`